h/api/models.py

```diff
@@ -166,28 +202,173 @@ def get_analysis(cls):
         return cls.__analysis__
 
     @property
+    def uri(self):
+        """Return this annotation's URI or an empty string.
+
+        The uri is escaped and safe to be rendered.
+
+        The uri is a Markup object so it won't be double-escaped.
+
+        """
+        uri_ = self.get("uri")
+        if uri_:
+
+            # Convert non-string URIs into strings.
+            # If the URI is already a unicode string this will do nothing.
+            # We're assuming that URI cannot be a byte string.
+            uri_ = unicode(uri_)
+
+            return jinja2.escape(uri_)
+        else:
+            return ""
+
+    @property
+    def filename(self):
+        """Return the filename of this annotation's document, or "".
+
+        If the annotated URI is a file:// URI then return the filename part
+        of it, otherwise return "".
+
+        The filename is escaped and safe to be rendered.
+
+        If it contains escaped characters then the filename will be a
+        Markup object so it won't be double-escaped.
+
+        """
+        if self.uri.lower().startswith("file:///"):
+            # self.uri is already escaped so we don't need to escape it again
```

Comment:
There's a bunch of places you've got comments like this. I don't think this is the right approach. The entire point of a taintedness approach to string escaping is that you don't need to examine the whole call stack to understand if what you're doing is safe. If this function returns something that needs to be escaped, it should escape it. Using `jinja2.escape(self.uri.split('/')[-1])` will not double escape anything.